From 6b98913afa6ce358c993f3f0d22e4a45b3d1750d Mon Sep 17 00:00:00 2001
From: Mikhail Lukashov <mlukashov@tabit.pro>
Date: Wed, 6 Oct 2021 15:05:22 +0300
Subject: [PATCH] qubes-core-qubesdb\daemon: fix build using VS2019

daemon/qubesdb_internal.h: error C2229: struct 'db_daemon_data' has an illegal zero-sized array
daemon/db-cmds.c: error C2065: 'ssize_t': undeclared identifier
---
 daemon/buffer.h           | 1 +
 daemon/db-cmds.c          | 6 +++++-
 daemon/db-daemon.c        | 2 +-
 daemon/qubesdb_internal.h | 8 ++++----
 include/qubesdb.h         | 2 +-
 5 files changed, 12 insertions(+), 7 deletions(-)

diff --git a/daemon/buffer.h b/daemon/buffer.h
index b41f0f9..83c9b50 100644
--- a/daemon/buffer.h
+++ b/daemon/buffer.h
@@ -19,6 +19,7 @@
  *
  */
 
+#pragma once
 #ifndef _BUFFER_H
 #define _BUFFER_H
 
diff --git a/daemon/db-cmds.c b/daemon/db-cmds.c
index ba2f15b..38c85ba 100644
--- a/daemon/db-cmds.c
+++ b/daemon/db-cmds.c
@@ -15,9 +15,13 @@
 
 #include <libvchan.h>
 
+#include "buffer.h"
 #include <qubesdb.h>
 #include "qubesdb_internal.h"
 
+#define size_t SIZE_T
+#define ssize_t SSIZE_T
+
 /** Check if given string matches path specification (i.e. have only allowed
  * characters). This function allows for '/' at the end, so if particular
  * command doesn't permit so, appropriate handle_* function should additionally
@@ -144,7 +148,7 @@ int vchan_write_nonblock(libvchan_t *vchan, char *buf, size_t size) {
 int write_vchan_or_client(struct db_daemon_data *d, struct client *c,
         char *data, int data_len) {
     int ret, count;
-    struct buffer *write_queue;
+    struct buffer *write_queue = NULL;
     int buf_datacount;
 
     if (c == NULL) {
diff --git a/daemon/db-daemon.c b/daemon/db-daemon.c
index aa660d0..2a9898a 100644
--- a/daemon/db-daemon.c
+++ b/daemon/db-daemon.c
@@ -9,7 +9,6 @@
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <signal.h>
-#include "buffer.h"
 #else
 #include <windows.h>
 #include <sddl.h>
@@ -35,6 +34,7 @@ struct thread_param {
 };
 #endif
 
+#include "buffer.h"
 #include <qubesdb.h>
 #include "qubesdb_internal.h"
 
diff --git a/daemon/qubesdb_internal.h b/daemon/qubesdb_internal.h
index 7b90b5b..bcca372 100644
--- a/daemon/qubesdb_internal.h
+++ b/daemon/qubesdb_internal.h
@@ -4,11 +4,11 @@
 
 #include <libvchan.h>
 
-#ifndef WIN32
 #include "buffer.h"
-#else
+#ifdef WIN32
 #include <log.h>
 #include <pipe-server.h>
+#include <qubesdb.h>
 #endif
 
 #ifndef WIN32
@@ -73,8 +73,6 @@ struct db_daemon_data {
     int remote_connected;       /* if remote daemon connected and ready for
                                  * processing requests (i.e. have
                                  * synchronised database */
-    struct buffer *vchan_buffer;/* vchan write buffer */
-    struct qdb_hdr vchan_pending_hdr; /* retrieved header but data not yet handled */
 #ifdef WIN32
     PIPE_SERVER pipe_server;
     SECURITY_ATTRIBUTES sa;
@@ -89,6 +87,8 @@ struct db_daemon_data {
     struct qubesdb *db;         /* database */
     int multiread_requested;    /* have requested multiread, if not - drop such
                                    responses */
+    struct buffer *vchan_buffer;/* vchan write buffer */
+    struct qdb_hdr vchan_pending_hdr; /* retrieved header but data not yet handled */
 };
 
 struct qubesdb *qubesdb_init(send_watch_notify_t);
diff --git a/include/qubesdb.h b/include/qubesdb.h
index e3af043..b53fc61 100644
--- a/include/qubesdb.h
+++ b/include/qubesdb.h
@@ -1,4 +1,4 @@
-
+#pragma once
 #include <stdint.h>
 
 /* Possible contexts:
-- 
2.33.0.windows.2

